site/content/entry/citra-progress-report-2017-september.md

```diff
@@ -0,0 +1,360 @@
+was written by [ds84182](https://github.com/ds84182) about two years ago, only to
+be abandoned due to not knowing how the configuration of them was done. The second
+attempt was written by [JayFoxRox](https://github.com/JayFoxRox), but was also
+abandoned for the same reason.
+
+But, after extensive research on geometry shaders was made by [fincs](https://github.com/fincs),
+the API was implemented in [ctrulib](https://github.com/smealum/ctrulib) and
+[citro3d](https://github.com/fincs/citro3d), and examples were written to demonstrate
+how to use it. Now that the community knew exactly how they worked, [wwylele](https://github.com/wwylele)
+picked up where [JayFoxRox](https://github.com/JayFoxRox) left off, cleaned up
+the code they wrote, and added the missing pieces.
+
+After almost three years, and three different attempts to make it work, Citra now
+has a full, complete, and correct implementation of geometry shaders!
+
+## [Implement custom clip plane](https://github.com/citra-emu/citra/pull/2900) by [wwylele](https://github.com/wwylele)
+
+After the geometry shader (or the vertex shader, if it wasn't enabled), the vertices
+are "assembled" into a collection of triangles. After *that*, to make rendering
+more efficient, the triangles are then compared to 6 planes that make up the cube
+in which objects are actually visible by the camera. Any triangles outside of that
+cube are deleted, and any triangles that are partially inside the cube are split
+by the sides of the cube, and the resulting triangle outside of the cube is also
+deleted.
+
+But the 3DS allows games to add a 7th plane whose position is fully customizable.
+Although no games are known to use this feature right now, because it is indeed
+a feature of the 3DS' GPU, and because implementing it was fairly straightforward,
+[wwylele](https://github.com/wwylele) decided to just go ahead and do it, just in
+case someone decided to use it for some reason.
+
+## [Optimized Morton](https://github.com/citra-emu/citra/pull/2951) by [huwpascoe](https://github.com/huwpascoe)
+
+Morton code is a function that interleaves multi-dimensional numbers into a one-dimensional
+number. Although it may seem like a very esoteric function, it's actually extremely
+useful in fields like linear algebra, databases, and what the 3DS uses it for:
+texture mapping.
+
+Computers have an intermediate chunk of memory between RAM and the CPU called a
+cache. Caches are seperated into lines, each of which can hold one data item. GPUs
+also have a cache, also seperated into lines. Because they are seperated like this,
+if a texture is loaded into the cache, it would have to span multiple cache lines,
+or even not fit into the cache completely, thus making transformations on it slow,
+as it would have to load and store pieces of it from RAM multiple times.
+
+To avoid this, GPUs can Morton encode textures so that two-dimensional manipulations
```

Comment:
Is it true that GPUs widely use this (I have no idea) ... I just assumed this was something somewhat specific to PICA

Reply:
[Yes](https://www.opengl.org/discussion_boards/showthread.php/176705-What-is-texture-swizzling), and [no](https://fgiesen.wordpress.com/2011/01/17/texture-tiling-and-swizzling/) respectively. In fact, it's so common, it's [built into OpenGL's texture handling since 3.3](https://www.khronos.org/opengl/wiki/Texture#Swizzle_mask).